LibreNMS/OS/SmOs.php

```diff
@@ -3,20 +3,97 @@
 namespace LibreNMS\OS;
 
 use LibreNMS\Device\WirelessSensor;
-use LibreNMS\Interfaces\Discovery\Sensors\WirelessRssiDiscovery;
 use LibreNMS\Interfaces\Discovery\Sensors\WirelessFrequencyDiscovery;
 use LibreNMS\Interfaces\Discovery\Sensors\WirelessMseDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRateDiscovery;
+use LibreNMS\Interfaces\Discovery\Sensors\WirelessRssiDiscovery;
 use LibreNMS\OS;
 
 class SmOs extends OS implements
+    WirelessRateDiscovery,
     WirelessRssiDiscovery,
     WirelessFrequencyDiscovery,
     WirelessMseDiscovery
 {
+    private $radioLabels;
+    private $linkLabels;
+
+    public function discoverWirelessRate()
+    {
+        $oids = snmpwalk_group($this->getDevice(), 'linkTxETHCapacity', 'SIAE-RADIO-SYSTEM-MIB', 2);
+        $oids = snmpwalk_group($this->getDevice(), 'linkRxETHCapacity', 'SIAE-RADIO-SYSTEM-MIB', 2, $oids);
+        $sensors = [];
+
+        foreach ($oids as $link => $radioEntry) {
+            $totalOids = ['rx' => [], 'tx' => []];
+
+            foreach ($radioEntry as $radio => $entry) {
+                $index = "$link.$radio";
+                if (isset($entry['linkTxETHCapacity'])) {
+                    $txOid = '.1.3.6.1.4.1.3373.1103.80.17.1.10.' . $index;
+                    $totalOids['tx'][] = $txOid;
+                    $sensors[] = new WirelessSensor(
+                        'rate',
+                        $this->getDeviceId(),
+                        $txOid,
+                        'tx',
+                        $index,
+                        $this->getLinkLabel($link) . ' Tx ' . $this->getRadioLabel($radio),
+                        $entry['linkTxETHCapacity'],
+                        1000
+                    );
+                }
+
+                if (isset($entry['linkRxETHCapacity'])) {
+                    $rxOid = '.1.3.6.1.4.1.3373.1103.80.17.1.11.' . $index;
+                    $totalOids['rx'][] = $rxOid;
+                    $sensors[] = new WirelessSensor(
+                        'rate',
+                        $this->getDeviceId(),
+                        $rxOid,
+                        'rx',
+                        $index,
+                        $this->getLinkLabel($link) . ' Rx ' . $this->getRadioLabel($radio),
+                        $entry['linkRxETHCapacity'],
+                        1000
+                    );
+                }
+            }
+
+            if (!empty($totalOids['rx'])) {
+                $sensors[] = new WirelessSensor(
+                    'rate',
+                    $this->getDeviceId(),
+                    $totalOids['rx'],
+                    'total-rx',
+                    $index,
```

Comment:
@murrant Sorry, hunting old potentiation issues here, `$index` comes from previous foreach loop, is this really correct?

Reply:
does not seem correct.